aten/src/ATen/function_wrapper.py

```diff
@@ -1188,11 +1304,52 @@ def gen_namespace_function(option, multidispatch_tensors):
                     option, native_arguments=option['native_actuals'])
 
             if is_factory_method:
-                fn_definition = C10_FACTORY_DEFINITION.substitute(
+                fn_definition = C10_FACTORY_DEFINITION_UNDERSCORE.substitute(
                     option, static_dispatch_function_body=static_dispatch_function_body)
             else:
                 fn_definition = C10_FUNCTION_DEFINITION.substitute(
                     option, static_dispatch_function_body=static_dispatch_function_body)
+
+            return FunctionCode(definition=fn_definition, declaration=fn_declaration)
+
+        def gen_collapsed_function_declaration(option):
+            declaration = DEPRECATED_COLLAPSED_FUNCTION_DECLARATION if option['deprecated'] else COLLAPSED_FUNCTION_DECLARATION
+            fn_declaration = declaration.substitute(option)
+
+            expanded_native_actuals = option['collapsed_native_actuals'][:]
+            index = expanded_native_actuals.index('options')
+            expanded_native_actuals.remove('options')
+            expanded_native_actuals.insert(index, 'options.pinned_memory()')
+            expanded_native_actuals.insert(index, 'options.device()')
+            
+             # special case for 'sparse_coo_tensor' and '_sparse_coo_tensor_unsafe'
+             # issue #30405
+            if (option['name'] == 'sparse_coo_tensor' or option['name'] == '_sparse_coo_tensor_unsafe'):
+                expanded_native_actuals.insert(index, 'options.layout_opt()')
+            else:
+                expanded_native_actuals.insert(index, 'options.layout()')
+            expanded_native_actuals.insert(index, 'typeMetaToScalarType(options.dtype())')
+
+            fn_definition = COLLAPSED_FACTORY_DEFINITION.substitute(option, expanded_native_actuals=expanded_native_actuals)
+            return FunctionCode(definition=fn_definition, declaration=fn_declaration)
+
+        def gen_native_dispatch_declaration(option):
+            declaration = NATIVE_DISPATCH_DECLARATION_CONST
+            fn_declaration = declaration.substitute(option, type_method_formals= collapse_formals(option['method_formals_with_defaults']))
+
+            expanded_native_actuals = option['collapsed_method_actuals'][:]
+            expanded_native_actuals.remove('const_cast<Tensor&>(*this)')
+            index = expanded_native_actuals.index('options')
+            expanded_native_actuals.remove('options')
+            expanded_native_actuals.insert(index, 'pin_memory')
+            expanded_native_actuals.insert(index, 'device')
+            expanded_native_actuals.insert(index, 'layout')
+            expanded_native_actuals.insert(index, 'dtype')
```

Comment:
This is, like, reverse order of what they're order in `expanded_native_actuals` in the end is, right? Consider using a slice assignment: `expanded_native_actuals[index:index] = ['dtype', 'layout', 'device', 'pin_memory']`